Public/Src/Engine/Processes/SharedOpaqueJournal.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.ContractsLight;
+using System.IO;
+using System.Linq;
+using BuildXL.Pips.Operations;
+using BuildXL.Utilities;
+using BuildXL.Utilities.Collections;
+
+namespace BuildXL.Processes
+{
+    /// <summary>
+    /// TODO 
+    /// </summary>
+    /// <remarks>
+    /// NOTE: not thread-safe
+    /// </remarks>
+    public sealed class SharedOpaqueJournal : IDisposable
+    {
+        private readonly PathTable m_pathTable;
+        private readonly AbsolutePath[] m_rootDirectories;
+        private readonly BuildXLWriter m_bxlWriter;
+
+        /// <summary>
+        /// Absolute path of this journal file.
+        /// </summary>
+        public string JournalPath { get; }
+
+        /// <summary>
+        /// Absolute path of the directory where this journal file is saved.
+        /// </summary>
+        public string JournalDirectory => Path.GetDirectoryName(JournalPath);
+
+        /// <summary>
+        /// Creates a journal for a given process.
+        /// 
+        /// Shared opaque directory outputs of <paramref name="process"/> are used as root directories and
+        /// <see cref="Pip.FormattedSemiStableHash"/> is used as journal base name.
+        ///
+        /// <seealso cref="SharedOpaqueJournal(PathTable, IEnumerable{AbsolutePath}, AbsolutePath)"/>
+        /// </summary>
+        public SharedOpaqueJournal(PipExecutionContext context, Process process, AbsolutePath journalDirectory)
+            : this(
+                  context.PathTable, 
+                  process.DirectoryOutputs.Where(d => d.IsSharedOpaque).Select(d => d.Path),
+                  GetJournalFileForProcess(context.PathTable, journalDirectory, process))
+        {
+            Contract.Requires(process != null);
+            Contract.Requires(context != null);
+            Contract.Requires(journalDirectory.IsValid);
+            Contract.Requires(Directory.Exists(journalDirectory.ToString(context.PathTable)));
+        }
+
+        /// <summary>
+        /// Creates a new journal.
+        /// </summary>
+        /// <param name="pathTable">Path table.</param>
+        /// <param name="rootDirectories">Only paths under one of the root directories are recorded in <see cref="RecordFileWrite(AbsolutePath)"/>.</param>
+        /// <param name="journalPath">File to which to write recorded accesses.</param>
+        public SharedOpaqueJournal(PathTable pathTable, IEnumerable<AbsolutePath> rootDirectories, AbsolutePath journalPath)
+        {
+            Contract.Requires(pathTable.IsValid);
+            Contract.Requires(rootDirectories != null);
+
+            m_pathTable = pathTable;
+            m_rootDirectories = rootDirectories.ToArray();

```

Comment:
>rootDirectories.ToArray(); [](start = 32, length = 26)

Maybe require an IReadOnlyCollection to begin with? So you avoid copies if the user of this constructor already has one at hand